src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs

```diff
@@ -25,39 +25,45 @@ internal static BoundBlock InjectInstrumentation(MethodSymbol method, BoundBlock
-                    FieldSymbol payloadField = GetPayloadField(method, methodOrdinal, payloadType, factory);
-                    bool methodHasExplicitBlock = MethodHasExplicitBlock(method);
+                    ArrayTypeSymbol modulePayloadType = ArrayTypeSymbol.CreateCSharpArray(compilation.Assembly, payloadType);
 
-                    // Synthesize the instrumentation and collect the spans of interest.
+                    // Create the symbol for the instrumentation payload.
+                    LocalSymbol methodPayload = factory.SynthesizedLocal(payloadType, kind: SynthesizedLocalKind.InstrumentationPayload, syntax: methodBody.Syntax);
 
+                    // Synthesize the instrumentation and collect the spans of interest.
                     // PROTOTYPE (https://github.com/dotnet/roslyn/issues/9819): Try to integrate instrumentation with lowering, to avoid an extra pass over the bound tree.
-                    BoundBlock newMethodBody = InstrumentationInjectionRewriter.InstrumentMethod(method, methodBody, methodHasExplicitBlock, payloadField, compilationState, diagnostics, debugDocumentProvider, out dynamicAnalysisSpans);
+                    BoundBlock newMethodBody = InstrumentationInjectionRewriter.InstrumentMethod(method, methodBody, methodHasExplicitBlock, methodPayload, compilationState, diagnostics, debugDocumentProvider, out dynamicAnalysisSpans);
 
                     // Synthesize the initialization of the instrumentation payload array, using concurrency-safe code:
                     //
-                    // if (payloadField == null)
-                    //     Instrumentation.CreatePayload(mvid, method, ref payloadField, payloadLength);
-                    
-                    BoundExpression mvid = factory.ModuleVersionId();
+                    // var payload = PID.PayloadRootField[methodIndex];
+                    // if (payload == null)
+                    //     payload = Instrumentation.CreatePayload(mvid, methodIndex, ref PID.PayloadRootField[methodIndex], payloadLength);
 
-                    BoundExpression methodToken = factory.MethodDefinitionToken(method);
-                    BoundStatement createPayloadCall = factory.ExpressionStatement(factory.Call(null, createPayload, mvid, methodToken, factory.Field(null, payloadField), factory.Literal(dynamicAnalysisSpans.Length)));
+                    BoundStatement payloadInitialization = factory.Assignment(factory.Local(methodPayload), factory.ArrayAccess(factory.InstrumentationPayloadRoot(analysisKind, modulePayloadType), ImmutableArray.Create(factory.MethodDefIndex(method))));
+                    BoundExpression mvid = factory.ModuleVersionId();
+                    BoundExpression methodToken = factory.MethodDefIndex(method);
+                    BoundExpression payloadSlot = factory.ArrayAccess(factory.InstrumentationPayloadRoot(analysisKind, modulePayloadType), ImmutableArray.Create(factory.MethodDefIndex(method)));
```

Comment:
I think that a bound tree is supposed to be a tree, not a graph, which precludes using a single node as an operand to more than one node.